plugin/pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -52,9 +88,66 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 		podPorts:                  schedutil.GetUsedPorts(pod),
 		matchingAntiAffinityTerms: matchingTerms,
 	}
-	for predicateName, precomputeFunc := range predicatePrecomputations {
+	for predicateName, precomputeFunc := range predicateMetadataProducers {
 		glog.V(10).Infof("Precompute: %v", predicateName)
 		precomputeFunc(predicateMetadata)
 	}
 	return predicateMetadata
 }
+
+// RemovePod changes predicateMetadata assuming that the given `deletedPod` is
+// deleted from the system.
+func (meta *predicateMetadata) RemovePod(deletedPod *v1.Pod) error {
+	if deletedPod.GetUID() == meta.pod.GetUID() {
+		return fmt.Errorf("deletedPod and meta.pod must not be the same.")
+	}
+	// Delete any anti-affinity rule from the deletedPod.
+	delete(meta.matchingAntiAffinityTerms, deletedPod.GetUID())
+	// All pods in the serviceAffinityMatchingPodList are in the same namespace.
+	// So, if the namespace of the first one is not the same as the namespace of the
+	// deletedPod, we don't need to check the list, as deletedPod isn't in the list.
+	if meta.serviceAffinityInUse &&
+		len(meta.serviceAffinityMatchingPodList) > 0 &&
+		deletedPod.Namespace == meta.serviceAffinityMatchingPodList[0].Namespace {
+		deletedPodIndex := -1
+		for i, pod := range meta.serviceAffinityMatchingPodList {
+			if pod.GetUID() == deletedPod.GetUID() {
+				deletedPodIndex = i
+				break
+			}
+		}
+		if deletedPodIndex >= 0 {
+			meta.serviceAffinityMatchingPodList = append(meta.serviceAffinityMatchingPodList[:deletedPodIndex], meta.serviceAffinityMatchingPodList[deletedPodIndex+1:]...)
+		}
+	}
+	return nil
+}
+
+// AddPod changes predicateMetadata assuming that `newPod` is added to the
+// system.
+func (meta *predicateMetadata) AddPod(addedPod *v1.Pod, nodeInfo *schedulercache.NodeInfo) error {
```

Comment:
We do need this. The preemption first removes all the lower priority pods and after adding the pending pod, tries to reprieve as many removed pods as possible. This method is useful when adding the pending pod and adding those removed pods back.